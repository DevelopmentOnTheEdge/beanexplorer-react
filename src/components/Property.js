import React         from 'react';
import PropTypes     from 'prop-types';
import classNames    from 'classnames';
import PropertyInput from './PropertyInput'


class Property extends React.Component
{
  getPath(){
    if(this.props.path){
      return this.props.path;
    }else{
      return this.props.bean.order[this.props.id];
    }
  }

  render() {
    const path = this.getPath();
    const meta  = this.props.bean.meta[path];
    const id    = path.substring(path.lastIndexOf("/")+1) + "PropertyInput";

    const label = <label htmlFor={id} className={meta.type === "Boolean" ? 'form-check-label' : 'form-control-label'}>
                         {meta.displayName || id}</label>;

    const messageElement = meta.message ? <small className={this.props.messageClassName || "form-control-feedback"}>{meta.message}</small> : undefined;

    let hasStatusClasses = classNames(
      {'has-error' : meta.status === 'error'},
      {'has-warning' : meta.status === 'warning'},
      {'has-success' : meta.status === 'success'},
    );

    const classNameForm = (meta.type === "Boolean")
      ? this.props.classNameFormCheck || 'form-check property'
      : this.props.classNameFormGroup || 'form-group property';

    const cssClasses = meta.cssClasses || 'col-lg-12';

    const outerClasses = classNames(
      cssClasses,
      {'display-none' : meta.hidden}
    );

    const classes = classNames(
      classNameForm,
      hasStatusClasses,
      {'required' : meta.canBeNull !== true}
    );

    if(this.props.inline)
    {
      if(meta.type === "Boolean")
      {
        return (
          <div className="form-check mb-2 mr-sm-2">
            <PropertyInput {...this.props} />
            {label}
          </div>
        );
      }
      else
      {
        return (
          <PropertyInput {...this.props} controlClassName="mb-2 mr-sm-2"/>
        );
      }
    }
    else
    {
      if(meta.type === "Boolean")
      {
        return (
          <div className={outerClasses}>
            <div className={classes}>
              <PropertyInput {...this.props} />
              {label}
              {messageElement}
            </div>
          </div>
        );
      }
      else if(meta.labelField)
      {
        return (
          <div className={classNames('form-group property property-label', meta.cssClasses || 'col-lg-12', hasStatusClasses)}>
            <PropertyInput {...this.props} />
          </div>
        );
      }
      else
      {
        return (
          <div className={outerClasses}>
            <div className={classes}>
              {label}
              <div className="controls">
                <PropertyInput {...this.props} />
                {messageElement}
              </div>
            </div>
          </div>
        );
      }
    }

  }

}

Property.defaultProps = {
  localization: {
    locale: 'en',
    clearAllText: 'Clear all',
    clearValueText: 'Clear value',
    noResultsText: 'No results found',
    searchPromptText: 'Type to search',
    placeholder: 'Select ...',
    loadingPlaceholder: 'Loading...'
  },
};

Property.propTypes = {
  bean: PropTypes.object.isRequired,
  path: PropTypes.string,
  id: PropTypes.number,
  inline: PropTypes.bool,
  onChange: PropTypes.func,
  localization: PropTypes.object
};

export default Property;
