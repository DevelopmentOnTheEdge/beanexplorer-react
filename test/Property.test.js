import React from 'react';
import Property from '../src/components/Property';
import {mount} from 'enzyme';
import bean from '../src/testJson.json';


test('get by id', () => {
  const handle = jest.fn();

  const wrapper = mount(
    <Property id={0} bean={bean} onChange={handle} value={"test"}/>
  );

  expect(wrapper.instance().getPath()).toEqual('/textInput');
});

test('get by path', () => {
  const handle = jest.fn();

  const wrapper = mount(
    <Property path={"/labelField"} bean={bean} onChange={handle} value={""}/>
  );

  expect(wrapper.instance().getPath()).toEqual('/labelField');
});
