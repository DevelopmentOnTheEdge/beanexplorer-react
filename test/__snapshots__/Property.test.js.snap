// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simple property 1`] = `
<div
  className="form-group property "
>
  <div
    className="input"
  >
    <label
      className={undefined}
      htmlFor="testNameField"
    >
      Combo Box
    </label>
    <div
      className="controls"
    >
      <select
        className="form-control"
        defaultValue="testValue"
        id="testNameField"
        onChange={[Function]}
      >
        <option
          value="bar"
        >
          foo
        </option>
        <option
          value="bar2"
        >
          foo 2
        </option>
      </select>
    </div>
  </div>
</div>
`;
